src/kpro.erl

```diff
@@ -333,14 +370,23 @@ decode(StructName, Bin) when is_atom(StructName) ->
   kpro_structs:decode(StructName, Bin).
 
 decode_message_stream(<<>>, Acc) ->
-  lists:reverse(Acc);
+  Acc;
 decode_message_stream(Bin, Acc) ->
   {Msg, Rest} =
     try decode(kpro_Message, Bin)
     catch error : {badmatch, _} ->
       {?incomplete_message, <<>>}
     end,
-  decode_message_stream(Rest, [Msg | Acc]).
+  NewAcc =
+    case Msg of
+      #kpro_Message{attributes = Attr} = Msg when ?KPRO_IS_GZIP_ATTR(Attr) ->
+        decode_message_stream(zlib:gunzip(Msg#kpro_Message.value), Acc);
+      #kpro_Message{attributes = Attr} = Msg when ?KPRO_IS_SNAPPY_ATTR(Attr) ->
+        decode_message_stream(java_snappy_unpack(Msg#kpro_Message.value), Acc);
```

Comment:
yes, kafka can handle raw compressed messages (without chunking).
and you are right, it decompresses messages, and compress back with offsets assigned, then write to disk.